File: two_sum.rb
# Given an array of integers, return indices of the two numbers such that they add up to a specific target.
#
# You may assume that each input would have exactly one solution, and you may not use the same element twice.
#
# Example:
#
# Given nums = [2, 7, 11, 15], target = 9,
#
# Because nums[0] + nums[1] = 2 + 7 = 9,
# return [0, 1].

# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    return [] if nums.size < 2

    lookup = {}

    for index in (0..nums.size - 1)
        return [index, lookup[target - nums[index]]] if lookup[target - nums[index]]

        lookup[nums[index]] = index
    end

    return []
end

File: twoSum.c
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        unordered_map<int, int> hash;
        vector<int> result;
        for (int i = 0; i < numbers.size(); i++) {
            int numberToFind = target - numbers[i];

//             if numberToFind is in map, return them
            if (hash.find(numberToFind) != hash.end()) {
//                 +` because indices are NOT zero based
                result.push_back(hash[numberToFind] + 1);
                return result;
            }

//             number not found.  Put it in the map.
            hash[numbers[i]] = i;
        }
        return result;
    }
};
